src/mrpro/algorithms/optimizers/pdhg.py

```diff
@@ -0,0 +1,204 @@
+"""Primal-Dual Hybrid Gradient Algorithm (PDHG)."""
+
+from __future__ import annotations
+
+import warnings
+from collections.abc import Callable, Sequence
+from dataclasses import dataclass
+
+import torch
+
+from mrpro.algorithms.optimizers.OptimizerStatus import OptimizerStatus
+from mrpro.operators import (
+    IdentityOp,
+    LinearOperator,
+    LinearOperatorMatrix,
+    ProximableFunctional,
+    ProximableFunctionalSeparableSum,
+)
+from mrpro.operators.functionals import ZeroFunctional
+
+
+@dataclass
+class PDHGStatus(OptimizerStatus):
+    """Status of the PDHG algorithm."""
+
+    objective: Callable[[*tuple[torch.Tensor, ...]], torch.Tensor]
+    dual_stepsize: float | torch.Tensor
+    primal_stepsize: float | torch.Tensor
+    relaxation: float | torch.Tensor
+    duals: Sequence[torch.Tensor]
+    relaxed: Sequence[torch.Tensor]
+
+
+def _norm_squared(*values: torch.Tensor) -> torch.Tensor:
+    """Calculate the squared L2 norm of the stack of tensors."""
+    return sum([torch.vdot(value.flatten(), value.flatten()).real for value in values], start=torch.tensor(0.0))
+
+
+def pdhg(
+    f: ProximableFunctionalSeparableSum | ProximableFunctional | None,
+    g: ProximableFunctionalSeparableSum | ProximableFunctional | None,
+    operator: LinearOperator | LinearOperatorMatrix | None,
+    initial_values: Sequence[torch.Tensor],
+    max_iterations: int = 32,
+    tolerance: float = 0,
+    primal_stepsize: float | None = None,
+    dual_stepsize: float | None = None,
+    relaxation: float = 1.0,
+    initial_relaxed: Sequence[torch.Tensor] | None = None,
+    initial_duals: Sequence[torch.Tensor] | None = None,
+    callback: Callable[[PDHGStatus], None] | None = None,
+) -> tuple[torch.Tensor, ...]:
+    r"""Primal-Dual Hybrid Gradient Algorithm (PDHG).
+
+    Solves the minimization problem
```

Comment:
Can we add some more math here and then also add the math description of the steps of the iteration further below in the code? Similar to the example script?